hail/src/main/scala/is/hail/linalg/BlockMatrix.scala

```diff
@@ -1546,6 +1441,62 @@ private class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix)
   @transient override val partitioner: Option[Partitioner] = Some(gp)
 }
 
+case class BlockMatrixRectanglesRDD(rectangles: Array[Array[Long]], bm: BlockMatrix)
+  extends RDD[(Int, BDM[Double])](bm.blocks.sparkContext, Nil) {
+
+  assert(rectangles.forall(rect => rect.length == 4))
+
+  val gp: GridPartitioner = bm.gp
+
+  override def compute(split: Partition, context: TaskContext): Iterator[(Int, BDM[Double])] = {
+    val rect = rectangles(split.index)
+    val rectStartRow = rect(0)
+    val rectEndRow = rect(1)
+    val rectStartCol = rect(2)
+    val rectEndCol = rect(3)
+
+    val rectData = new BDM[Double]((rectEndRow - rectStartRow).toInt, (rectEndCol - rectStartCol).toInt)
+    val blocksInRectangle = gp.rectangleBlocks(rect)
+    blocksInRectangle.foreach( blockIdx => {
+      val (blockRowIdx, blockColIdx) = gp.blockCoordinates(blockIdx)
+      val blockStartRow = blockRowIdx * gp.blockSize
+      val blockStartCol = blockColIdx * gp.blockSize
+      val blockEndRow = blockStartRow + gp.blockRowNRows(blockRowIdx)
+      val blockEndCol = blockStartCol + gp.blockColNCols(blockColIdx)
+
+      val blockRowSlice = overlapBlockSlice(rectStartRow, rectEndRow, blockStartRow, blockEndRow)
+      val blockColSlice = overlapBlockSlice(rectStartCol, rectEndCol, blockStartCol, blockEndCol)
+      val rectRowSlice = overlapRectSlice(rectStartRow, rectEndRow, blockStartRow, blockEndRow)
+      val rectColSlice = overlapRectSlice(rectStartCol, rectEndCol, blockStartCol, blockEndCol)
+
+      val block = BlockMatrix.block(bm, bm.blocks.partitions, gp, context, blockRowIdx, blockColIdx).get
+      rectData(rectRowSlice, rectColSlice) := block(blockRowSlice, blockColSlice)
```

Comment:
if the block is `None`, that's a sparse, all-zero block, right? That's not an error, we should just ignore it (since BDM defaults to zeros anyway).

Reply:
Ah I see. I thought because I asked for the blocks before-hand we would be guaranteed it exists, but that does not account for sparseness.